[Threading] Always round up when converting std::chrono::durations.

al45tair opened the pull request:
We were inadvertently rounding down, which was fine almost 100% of the time, but on Windows where the timer resolution is only 1ms, it made the test _very slightly_ flaky because we were asking to wait for 9ms rather than 10ms, then checking we'd waited at least 10ms.

We should explicitly round up, for that reason.

rdar://100236038

al45tair committed:
[Threading] Always round up when converting std::chrono::durations.

We were inadvertently rounding down, which was fine almost 100% of the
time, but on Windows where the timer resolution is only 1ms, it made
the test _very slightly_ flaky because we were asking to wait for
9ms rather than 10ms, then checking we'd waited at least 10ms.

We should explicitly round up, for that reason.

rdar://100236038

al45tair commented:
@swift-ci Please test

al45tair commented:
Hmmm. Seems this wasn't the only problem on Windows making the test flaky. I wonder if the server applied a time adjustment during the `SleepConditionVariableSRW()` call. Maybe the thing to do is to change the test to allow things to return 10% "sooner" than it thinks they should?

al45tair committed:
[UnitTests][Threading] Tweak ConditionVariable test for Windows.

Windows' behaviour wrt `Sleep` family functions can be odd.  Apparently,
if you specify a time lower than a system tick, they can return early.
This makes the test flaky when it shouldn't be.  To fix it, we use the
Multimedia functions to adjust the system tick count as low as we can
get it.

rdar://100236038

al45tair commented:
No, I don't think the time adjustment explanation is the right one. That _shouldn't_ really be happening because we're using `steady_clock`. I think it might be the odd behaviour of `Sleep` functions on Windows, where (apparently) if you try to wait for less than a system tick (by default 1/64s), they can return _early_.

Let's try increasing the system tick rate during the test. I'll run the tests a few times to make sure it works.

al45tair commented:
@swift-ci Please test Windows platform

mikeash reviewed (approved):
LGTM, once it works.

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

compnerd commented on include/swift/Threading/Impl/chrono.h:
```diff
@@ -0,0 +1,71 @@
+//===--- chrono.h - Utility functions for duration/time_point - -*- C++ -*-===//
```
Does it make sense to rename this header to `chrono_utils.h` to match the namespace and to differentiate it from the C++ header `chrono`?

al45tair committed:
[Threading] Rename chrono.h to match the namespace name.

rdar://100236038

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Threading][Win32] Fix the wait functions for Win32 to always over-wait.

Rather than bodging the test to make it more robust, fix the functions
in the Threading layer to behave the same as they do on other platforms,
i.e. to guarantee that they always wait *at least* the amount of time
you asked for.

rdar://100236038

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
Woo! That seems to have worked.

al45tair commented:
@swift-ci please smoke test
